### PR TITLE
do not block finality for "disabled" disputes

### PR DESCRIPTION
**Pull request by ordian**
- [x] test with zombienet-sdk
- [x] prdoc

Relevant Issues:
https://github.com/paritytech/polkadot-sdk/issues/3314 (connected to the cause)
https://github.com/paritytech/polkadot-sdk/issues/3345 (solves)

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5235389

File: polkadot/node/core/dispute-coordinator/src/import.rs [hunk 1]
```diff
@@ -584,7 +597,7 @@ impl ImportResult {
 			}
 		}
 
-		let new_state = CandidateVoteState::new(votes, env, now);
+		let new_state = CandidateVoteState::new(votes, env, now, is_disabled);
```
Comment (eskimor) [inline polkadot/node/core/dispute-coordinator/src/import.rs:600]: Why don't we make this part of the environment?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5235684

File: polkadot/node/core/dispute-coordinator/src/tests.rs [hunk 2]
```diff
@@ -2968,17 +2981,23 @@ fn participation_with_offchain_disabling() {
 			// let's disable validators 3, 4 on chain, but this should not affect this import
 			let disabled_validators = vec![ValidatorIndex(3), ValidatorIndex(4)];
 			handle_disabled_validators_queries(&mut virtual_overseer, disabled_validators).await;
```
Comment (eskimor) [inline polkadot/node/core/dispute-coordinator/src/tests.rs:2983]: I am confused. We are disabling validators 3 and 4, but the votes are coming from 1 and 2? Ok, so this should not affect the import, make sense, but why should it affect participation further down?

File: polkadot/node/core/dispute-coordinator/src/import.rs (hunk 1 above)
Comment (ordian) [reply polkadot/node/core/dispute-coordinator/src/import.rs:600]: you mean the `is_disabled` closure that looks up the global state of dispute coordinator?

**Review by eskimor — APPROVED** [review]

**Review by tdimitrov — APPROVED** [review]

File: polkadot/node/core/dispute-coordinator/src/import.rs [hunk 3]
```diff
@@ -237,13 +255,19 @@ impl CandidateVoteState<CandidateVotes> {
...
+		// We have a dispute, if we have votes on both sides, with at least one invalid vote
+		// from non-disabled validator or with votes on both sides and confirmed.
+		let has_non_disabled_invalid_votes =
+			votes.invalid.keys().any(|i| !env.disabled_indices().contains(i));
+		let byzantine_threshold = polkadot_primitives::byzantine_threshold(n_validators);
+		let votes_on_both_sides = !votes.valid.raw().is_empty() && !votes.invalid.is_empty();
+		let is_confirmed =
+			votes_on_both_sides && (votes.voted_indices().len() > byzantine_threshold);
+		let is_disputed =
+			is_confirmed || (has_non_disabled_invalid_votes && !votes.valid.raw().is_empty());
```
Comment (rphmeier) [inline polkadot/node/core/dispute-coordinator/src/import.rs:267]: is it worth building in _some_ safety margin here? or do we fully rest on the approval-checking protocol and assume that a non-disabled validator will raise a dispute before an approval credential is generated?

Comment (rphmeier) [reply polkadot/node/core/dispute-coordinator/src/import.rs:267]: Yes, this explanation holds water in theory, especially as a no-show. I am convinced this is fine to leave as-is, thanks.

**Review by rphmeier — APPROVED** [review]
Looks good

File: prdoc/pr_3358.prdoc [hunk 4]
```diff
@@ -0,0 +1,10 @@
+title: Do not stall finality on spam disputes
+
+doc:
+  - audience: Node Operator
+    description: |
+      This PR fixes the issue that periodically caused
+      finality stalls on Kusama due to disputes happening
+      there in combination with disputes spam protection mechanism.
```
[the remark this answers is not in the document]
Comment (bkchr) [reply prdoc/pr_3358.prdoc:8]: The tool you are searching for is called git blame 😬

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/finality-stall-on-kusama-15-02-2024-post-mortem/6398/1

**Pull request merged**